### PR TITLE
UNR-4464: fix flaky tests

### PR DESCRIPTION
**Pull request by thomas-yung**
#### Description
When `LayeredLBStrategyTest` is ran before `LocalDeploymentManager`, the latter fails due to some leftover state from the former, indicating that `LayeredLBStrategyTest` does not clean itself up properly 

To clean up:
  - Stop unnecessary deployments
  - Call `GEditor->RequestEndPlayMap();` (thanks @mironec 😄)

#### Tests
Run `LayeredLBStrategyTest` before `LocalDeploymentManager` (a few times, to check flakiness) - should pass

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp [hunk 1]
```diff
@@ -204,11 +205,17 @@ bool FCleanup::Update()
...
 	}
 
+	GEditor->RequestEndPlayMap();
+
 	return true;
 }
 
 LAYEREDLBSTRATEGY_TEST(GIVEN_strat_is_not_ready_WHEN_local_virtual_worker_id_is_set_THEN_is_ready)
 {
+	const auto Settings = GetMutableDefault<USpatialGDKEditorSettings>();
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:215]: Why is this required in these tests but not `OwnershipLockingPolicyTest`?

Comment (thomas-yung) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:215]: Mainly because `OwnershipLockingPolicyTest` was not recognised as flaky, but I can have a look to see if it requires similar changes 😄

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp [hunk 2]
```diff
@@ -220,13 +227,20 @@ LAYEREDLBSTRATEGY_TEST(GIVEN_strat_is_not_ready_WHEN_local_virtual_worker_id_is_
...
+
+	Settings->bAutoStartLocalDeployment = OldBAutoStartLocalDeployment;
 
 	return true;
 }
 
 LAYEREDLBSTRATEGY_TEST(
 	GIVEN_layered_strat_of_two_by_four_grid_strat_singleton_strat_and_default_strat_WHEN_get_minimum_required_workers_called_THEN_ten_returned)
 {
+	const auto Settings = GetMutableDefault<USpatialGDKEditorSettings>();
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:240]: This would be a good candidate for a test fixture (or some other form of refactoring).
Have a look at `SchemaTestFixture`, which is used for our schema tests.
I think we could bundle the setting changes into it at least, maybe even the AutomationOpenMap and FCleanup.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp [hunk 3]
```diff
@@ -238,13 +252,20 @@ LAYEREDLBSTRATEGY_TEST(
 	ADD_LATENT_AUTOMATION_COMMAND(FWaitForWorld(Data));
 	ADD_LATENT_AUTOMATION_COMMAND(FCreateStrategy(Data, MultiWorkerSettings, {}));
 	ADD_LATENT_AUTOMATION_COMMAND(FCheckMinimumWorkers(Data, this, 10));
+	ADD_LATENT_AUTOMATION_COMMAND(FCleanup(Data));
+
+	Settings->bAutoStartLocalDeployment = OldBAutoStartLocalDeployment;
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:257]: Beware that this will actually run right after AutomationOpenMap. Which is probably fine, since the map would have been opened with the correct setting by then, but it's worth keeping in mind, and may be better to reorder these things, so the order of operations is clearer.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:215]: Seconding Danny, would be good to look at the other tests which utilize AutomationOpenMap in the same pass.

File: CHANGELOG.md [hunk 4]
```diff
@@ -69,6 +69,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Fixed a bug which dropped component updates on authority delegation.
 - The DeploymentLauncher checks the validity of the simulated players deployment name.
 - Worker configuration watcher only rebuilds worker configs when `*.worker.json` files are changed.
+- Fixed side-effecting tests.
```
Comment (mironec) [inline CHANGELOG.md:72]: I would argue this is mostly an internal change and hence wouldn't need a release note, but I am bad at writing docs, so don't listen to me.

Comment (jessicafalk) [reply CHANGELOG.md:72]: would agree with that.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp (hunk 2 above)
Comment (thomas-yung) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:240]: Reordering the statements as you mentioned in a different comment makes this pretty simple to refactor into a `SetUp` method :)

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp [hunk 5]
```diff
@@ -348,3 +348,12 @@ GRIDBASEDLBSTRATEGY_TEST(GIVEN_two_cells_WHEN_actor_in_one_cell_THEN_strategy_re
 
 	return true;
 }
+
+GRIDBASEDLBSTRATEGY_TEST(zz_CleanUp_GridBasedLBStrategyTest)
+{
+	// Test are ran in alphabetical order and we don't want to make the tests slow by stopping the deployment after each
```
Comment (jessicafalk) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp:354]:
```suggestion
	// Tests are run in alphabetical order and we don't want to make the tests slow by stopping the deployment after each
```
or is it "ran"?

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/OwnershipLockingPolicy/OwnershipLockingPolicyTest.cpp [hunk 6]
```diff
@@ -1165,3 +1165,12 @@ OWNERSHIPLOCKINGPOLICY_TEST(
 
 	return true;
 }
+
+OWNERSHIPLOCKINGPOLICY_TEST(zz_CleanUp_OwnershipLockingPolicyTest)
```
Comment (thomas-yung) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/OwnershipLockingPolicy/OwnershipLockingPolicyTest.cpp:1169]: Very much open to suggestions on this

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/OwnershipLockingPolicy/OwnershipLockingPolicyTest.cpp [hunk 7]
```diff
@@ -1165,3 +1165,12 @@ OWNERSHIPLOCKINGPOLICY_TEST(
...
+{
+	// Test are ran in alphabetical order and we don't want to make the tests slow by stopping the deployment after each
```
Comment (jessicafalk) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/OwnershipLockingPolicy/OwnershipLockingPolicyTest.cpp:1171]:
```suggestion
	// Tests are run in alphabetical order and we don't want to make the tests slow by stopping the deployment after each
```

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp [hunk 8]
```diff
@@ -1273,7 +1273,8 @@ void FSpatialGDKEditorToolbarModule::OnAutoStartLocalDeploymentChanged()
 		{
 			// Bind the TryStartSpatialDeployment delegate if autostart is enabled.
 			UEditorEngine::TryStartSpatialDeployment.BindLambda([this](FString ForceSnapshot) {
-				if (GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking())
+				if (GetDefault<USpatialGDKEditorSettings>()->bAutoStartLocalDeployment
+					&& GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking())
```
Comment (jessicafalk) [inline SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp:1277]: I think that with this extra check in here we could now make this logic a bit easier and simply bind it once instead of binding and unbinding whenever the settings change. @improbable-valentyn  @mironec   what do you think?

though that could be done in a seaprate PR. the main thing we wanted to fix has been fixed 🎉

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp [hunk 9]
```diff
@@ -46,6 +47,18 @@ UWorld* GetAnyGameWorld()
...
+// Disable local deployments from starting and open map
+void SetUp()
+{
+	const auto Settings = GetMutableDefault<USpatialGDKEditorSettings>();
+	const auto OldBAutoStartLocalDeployment = Settings->bAutoStartLocalDeployment;
+	Settings->bAutoStartLocalDeployment = false;
+
+	AutomationOpenMap(SpatialConstants::EMPTY_TEST_MAP_PATH);
+
+	Settings->bAutoStartLocalDeployment = OldBAutoStartLocalDeployment;
```
Comment (jessicafalk) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:59]: ah nice. so as soon as we open the map, we can reset the value to its default? is `AutomationOpenMap` guaranteed to block until the map is open ?

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp [hunk 10]
```diff
@@ -348,3 +348,12 @@ GRIDBASEDLBSTRATEGY_TEST(GIVEN_two_cells_WHEN_actor_in_one_cell_THEN_strategy_re
 
 	return true;
 }
+
+GRIDBASEDLBSTRATEGY_TEST(zz_CleanUp_GridBasedLBStrategyTest)
```
Comment (jessicafalk) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp:352]: haha, using `zz` to make sure it's run last?

do we need the deployment up and running for these tests?

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp (hunk 5 above)
Comment (thomas-yung) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp:354]: that's my english AS level down the drain 😅

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp (hunk 10 above)
Comment (thomas-yung) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp:352]: zz to be doubly sure it's last 😄 
I checked with these this and Owenership...Test and they both require the deployment

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp (hunk 9 above)
Comment (thomas-yung) [reply SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:59]: `AutomationOpenMap` doesn't block itself, just triggers the start of a deployment (if enabled)
The tests then `FWaitForWorld` once the data is prepared

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp (hunk 8 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp:1277]: I did find it weird that we bind and unbind everytime...

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp [hunk 11]
```diff
@@ -46,6 +47,19 @@ UWorld* GetAnyGameWorld()
...
+bool SetUp()
+{
+	const auto Settings = GetMutableDefault<USpatialGDKEditorSettings>();
+	const auto OldBAutoStartLocalDeployment = Settings->bAutoStartLocalDeployment;
+	Settings->bAutoStartLocalDeployment = false;
+	UE_LOG(LogTemp, Warning, TEXT("Settings->bAutoStartLocalDeployment set to %d"), Settings->bAutoStartLocalDeployment)
+
+	AutomationOpenMap(SpatialConstants::EMPTY_TEST_MAP_PATH);
+
+	return OldBAutoStartLocalDeployment;
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp:60]: If this only applies to AutomationOpenMap it could be set back to the previous value here?

**Review by jessicafalk — APPROVED** [review]
🎉

**Review by improbable-danny — APPROVED** [review]
Looks good but have some reservations about the map persisting over multiple tests and requiring a test to shut that state down.

I don't think all of these tests require a map so spinning up/down a map for the ones which do might not be so bad (maybe)
